### ISSUE TITLE
Sniff to have one attribute per line

### ISSUE THREAD

**Issue by michnovka** [slevomat/coding-standard#1442]
```php
#[Column(unique: true)] protected string $name;
#[Cache] #[ManyToOne(targetEntity: 'City', inversedBy: 'attractions')] #[JoinColumn(name: 'city_id', referencedColumnName: 'id')] protected City $city;
```

would be good to be able to enforce

```php
#[Column(unique: true)]
protected string $name;

#[Cache] 
#[ManyToOne(targetEntity: 'City', inversedBy: 'attractions')] 
#[JoinColumn(name: 'city_id', referencedColumnName: 'id')] 
protected City $city;
```

attributes can be used also inside of function calls, constructor promotion etc.

Would be nice to combine with https://github.com/slevomat/coding-standard/issues/1259 in order to allow alphabetical sorting of attributes.

Maybe have an option to allow inline attributes in case that there is just one? can be useful for function params:

```php
function passwordHash(#[\SensitiveParameter] string $password) {
    var_dump(debug_backtrace());
}
```

**Comment by michnovka** [comment]
Willing to work on this, but lets first agree on the basic principles:

1. I would use SlevomatCodingStandard.Classes.AttributeSpacing
2. one option: ignoreForSingleAttribute (default false)
3. for now I would not allow specific cases depending on attribute target, either use everywhere or nowhere

**Comment by kukulich** [comment]
I'm not sure if `AttributeSpacingSniff` is right. This name may be preserved for sniff checking spaces between attribute and property/method/etc.

So maybe `RequireOneAttributeInLineSniff` (similar to `RequireOneNamespaceInFileSniff`) ?

And I would vote for new "namespace" `Attributes`.

**Comment by michnovka** [comment]
I think the Attributes namespace is handy and I also think the option to `ignoreForSingleAttribute` is useful.

However the name may be a bit misleading.

as look:

```php
#[Cache]
#[ManyToOne(targetEntity: 'City', inversedBy: 'attractions')]
#[JoinColumn(name: 'city_id', referencedColumnName: 'id')] protected City $city;
```

does satisfy the condition of "one attribute per line" but its clearly not what was intended.

The [SlevomatCodingStandard.Classes.ConstantSpacing](https://github.com/slevomat/coding-standard#slevomatcodingstandardclassesconstantspacing-) is also about lines, not spaces and so is `SlevomatCodingStandard.Classes.MethodSpacing` and `SlevomatCodingStandard.Classes.PropertySpacing`

Therefore I still vote for 
`SlevomatCodingStandard.Attributes.AttributeSpacing` or `SlevomatCodingStandard.Attributes.AttributeLineSpacing`

What do you think @kukulich ?

**Comment by kukulich** [comment]
Will think about it again...

**Comment by kukulich** [comment]
Ok, `SlevomatCodingStandard.Attributes.AttributeSpacing` with options:

`linesCountBetweenAttributes` - default `1` - will split attributes to more lines (yes, even `#[Cache, ManyToOne]`. `0` will join attributes to one line.
`linesCountBeforeOwner` (or Member or something else?) - default `0`
`singleAttributeOnOwnerLine` - default `false`

Do I forgot something?

**Comment by michnovka** [comment]
I would distinguish between merging attributes into one `#[]` block and line spacing.

so lets keep the syntax for `SlevomatCodingStandard.Attributes.AttributeSpacing` as you defined and add one two more sniffs too `SlevomatCodingStandard.Attributes.RequireAttributeJoining` and  `SlevomatCodingStandard.Attributes.DisallowAttributeJoining`

If set, these will either force all attributes to be in one `#[]` block or force separation into separate `#[]`. Then if split, in combination with `SlevomatCodingStandard.Attributes.AttributeSpacing` they can be also separated into multiple lines.

WDYT? @kukulich

**Comment by kukulich** [comment]
I'm ok with `DisallowAttributeJoining`, do we need `RequireAttributeJoining` ?

**Comment by michnovka** [comment]
Agreed, if somebody needs it, can be implemented later.

One more thing I want to ask:

```php
#[Column(unique: true)]
protected string $name;

#[Cache] 
#[ManyToOne(targetEntity: 'City', inversedBy: 'attractions')] 
#[JoinColumn(name: 'city_id', referencedColumnName: 'id')] 
protected City $city;
```

What specifies the number of lines between `protected string $name;` and `#[Cache]`?

```php
protected string $name;
#[Cache] 
```
or
```php
protected string $name;




#[Cache] 
```

**Comment by kukulich** [comment]
@michnovka Yes, `PropertySpacing` is about spacing between properties but `AttributeSpacing` should solve spacing between property and its attributes - that's what I meant with `linesCountBeforeOwner` option. However it may be added later or we may add special sniff for it.

**Comment by kukulich** [comment]
So I suggest this:

1. `SlevomatCodingStandard.Attributes.AttributeSpacing` with `linesCountBetweenAttributes`
2. `SlevomatCodingStandard.Attributes.DisallowAttributeJoining`
3. Solve `linesCountBeforeOwner` and `singleAttributeOnOwnerLine` in `AttributeSpacing` or new sniff, eg. `AttributeOwnerSpacing`

**Comment by michnovka** [comment]
This is my work in progress - https://github.com/slevomat/coding-standard/pull/1447 I have finished `DisallowAttributeJoiningSniff` and will look into the spacing now.

However I have some questions, which Id like to discuss under the PR. @kukulich can you please link this Issue with the PR and well resume conversation in there? Thanks

**Comment by michnovka** [comment]
so I am returning to discuss the `SlevomatCodingStandard.Attributes.AttributeSpacing`

My current implementation - https://github.com/slevomat/coding-standard/pull/1449 targets the core issue which I was trying to solve originally -> DO not allow attributes to be on one line
```php
#[Cache] #[ManyToOne(targetEntity: 'City', inversedBy: 'attractions')] #[JoinColumn(name: 'city_id', referencedColumnName: 'id')] protected City $city;
```
and instead force one attribute per line
```php
#[Cache] 
#[ManyToOne(targetEntity: 'City', inversedBy: 'attractions')] 
#[JoinColumn(name: 'city_id', referencedColumnName: 'id')] 
protected City $city;
```

I decided to not implement `linesCountBetweenAttributes` because I cannot imagine a case where it would be desirable to have arbitrary number of lines per between attributes.

When it comes to actual spacing, I think this should be addressed with `SlevomatCodingStandard.Classes.ConstantSpacing`, `SlevomatCodingStandard.Classes.PropertySpacing` etc. These sniffs take into account PHP comments, but these are getting replaced by attributes in all modern projects (symfony, doctrine, laravel, apiplatform,...). I think it makes no sense to try to force line spacing based on *attributes*, same as it does not have sense to force spacing based on *comments*.

Instead, we should expand the syntax of the other sniffs and add `minLinesCountBeforeWithAttribute` and `maxLinesCountBeforeWithAttribute`. This brings up a complication though, as we now have 4 types of properties/functions etc.

1. no comment and no attribute
2. has comment but no attribute
3. no comment but has attribute
4. has both comment and attributes

On top of that, there might be nightmarish code like

```php
// multiple attributes applicable
#[Printable] // allow printable inline comment
/** shoud be downloadable
    or some other long comment
*/
#[Downloadable()]
/** @var string $firstName store first name of user */
private string $firstname;
```

How do we handle mixing of comments and attributes? My first idea was to make sniffs enforcing order of attributes and comments. But that has no sense, since they can be interlaced.

##Proposed solution

1. I suggest we take my current implementation and rename the sniff to `SlevomatCodingStandard.Attributes.RequireAttributeOnDedicatedLine`. That will be the most useful sniff for most cases
2.  We should reconsider parameters to all line spacing sniffs and I suggest **we treat attributes as comments**. For BC I would keep the old params like `minLinesCountBeforeWithComment` but I would introduce new ones: `minLinesCountBeforeWithMetadata` where by **Metadata** I would consider both phpdocs and attributes. This will allow us to specify line spacing based on token type (property/function/method) and not just globally for all attributes.

Awaiting your opinion

**Comment by kukulich** [comment]
> I decided to not implement linesCountBetweenAttributes because I cannot imagine a case where it would be desirable to have arbitrary number of lines per between attributes.

Yes, I cannot imagine it either however I was already suprised before with users needs. That's why almost all sniffs have options.

> When it comes to actual spacing, I think this should be addressed with SlevomatCodingStandard.Classes.ConstantSpacing, SlevomatCodingStandard.Classes.PropertySpacing

These sniffs require number of lines between two constants or two properties. The sniffs cannot solve number of lines between attribute and its constant/property.

> How do we handle mixing of comments and attributes? My first idea was to make sniffs enforcing order of attributes and comments. But that has no sense, since they can be interlaced.

Yes, there should be special sniff for order between comments and attributes. Probably something like `RequireDocCommentBeforeAttributes` : https://github.com/slevomat/coding-standard/issues/1264

> I suggest we take my current implementation and rename the sniff to SlevomatCodingStandard.Attributes.RequireAttributeOnDedicatedLine

I'm willing to merge sniff as `DisallowMultipleAttributesPerLineSniff` (because we already have `MultipleUsesPerLineSniff`)  but only without the `ignoreForSingleAttribute` option. I don't like it and it's useless without fixer. However I don't think it should be part of this sniff.

**Comment by kukulich** [comment]
Implemented in https://github.com/slevomat/coding-standard/pull/1449
